### DISCUSSION TITLE
Refactoring of cohort and patch subroutines

### DISCUSSION THREAD

**Discussion by adrifoster** [NGEET/fates#1016] [Ideas]
I am working on some refactoring of the FATES derived types (i.e. `site`, `patch`, and `cohort` types). See my WIP branch [here](https://github.com/adrifoster/fates/tree/unit_test_framework).

I am planning on moving some of the methods associated with patches and cohorts into modules for those types. Right now, in the modules [`EDCohortDynamicsMod`](https://github.com/NGEET/fates/blob/main/biogeochem/EDCohortDynamicsMod.F90) and [`EDPatchDynamicsMod`](https://github.com/NGEET/fates/blob/main/biogeochem/EDPatchDynamicsMod.F90) there seem to be subroutines that should belong cohorts or patches (or sites).

@glemieux suggested I put them up here with my thoughts so others can chime in:

@ckoven, @rgknox, @rosiealice, @billsacks, @mpaiao, @JessicaNeedham, @jkshuman 

In EDCohortDynamicsMod (for now just focusing on these b/c there are a lot...):
1. [`create_cohort`](https://github.com/NGEET/fates/blob/5bb6e3e643b547e3c0dadcbe065e73ffa21b1053/biogeochem/EDCohortDynamicsMod.F90#L153)
 - this subroutine allocates memory for a new cohort, sets initial values for all cohort attributes, and then inserts the cohort into the correct place in the patch's cohort list 
 - [ this has been modified by me](https://github.com/adrifoster/fates/blob/3a391a72826423c83315687a4ffbdc0e56d6a75f/main/FatesCohortMod.F90#L517) so that the cohort-relevant methods (i.e. setting init values) are in their own cohort methods
 - *the insertion into the patch's linked list part seems like a `patch` method to me*
 - *specifically, if we make the linked list parts actual methods, we can make more easily test them*
 2. [`InitPRTObject`](https://github.com/NGEET/fates/blob/5bb6e3e643b547e3c0dadcbe065e73ffa21b1053/biogeochem/EDCohortDynamicsMod.F90#L466)
 - this subroutine initializes the PRT object to be associated with each cohort
 - there are no cohort objects used inside the subroutine
 - *this seems like a `PRT` method to me*
 3. [`terminate_cohorts`](https://github.com/NGEET/fates/blob/5bb6e3e643b547e3c0dadcbe065e73ffa21b1053/biogeochem/EDCohortDynamicsMod.F90#L737)
 - this subroutine loops through all the cohorts in a patch, checks to see which ones need to be deleted for various reasons, and then calls `terminate_cohort` (see below) and deallocates memory for each one marked for deletion
 - *this seems like a `patch` method (definitely NOT at `cohort` method) to me, or perhaps we should put it in some other kind of module*
 4. [`terminate_cohort`](https://github.com/NGEET/fates/blob/5bb6e3e643b547e3c0dadcbe065e73ffa21b1053/biogeochem/EDCohortDynamicsMod.F90#L854)
 - this subroutine puts the litter from a cohort in the appropriate places on a site/patch and updates the C flux for the site, removes the cohort from the patch's linked list, and then calls `DeallocateCohort` on that cohort (see below)
 - *seems like mostly `patch` methods, or perhaps in some other module*
5. [`SendCohortToLitter`](https://github.com/NGEET/fates/blob/5bb6e3e643b547e3c0dadcbe065e73ffa21b1053/biogeochem/EDCohortDynamicsMod.F90#L950)
- this subroutine transfers existing mass in all pools and all elements from a cohort to the appropriate litter pools
- needs access to the patch's `litter` pointer and the site's `flux_diags` pointer
6. [`DeallocateCohort`](https://github.com/NGEET/fates/blob/5bb6e3e643b547e3c0dadcbe065e73ffa21b1053/biogeochem/EDCohortDynamicsMod.F90#L1082)
- deallocates dynamic memory and objects inside the cohort structure
- *seems like a `cohort` method to me*
7. [`fuse_cohorts`](https://github.com/NGEET/fates/blob/5bb6e3e643b547e3c0dadcbe065e73ffa21b1053/biogeochem/EDCohortDynamicsMod.F90#L1111)
- this subroutine joins similar cohorts within a patch's cohort list
- this involves looping through a patch's cohort list (`patch` methods?) and
- rescaling the attributes of the new fused cohort (`cohort` methods?)
8. [`sort_cohorts`](https://github.com/NGEET/fates/blob/5bb6e3e643b547e3c0dadcbe065e73ffa21b1053/biogeochem/EDCohortDynamicsMod.F90#L1686)
- sorts the cohorts within a patch's linked list into the correct order
- *seems like `patch` methods to me*
9. [`insert_cohort`](https://github.com/NGEET/fates/blob/5bb6e3e643b547e3c0dadcbe065e73ffa21b1053/biogeochem/EDCohortDynamicsMod.F90#L1736)
- inserts a cohort into a patch's linked list
- *seems like it should be `patch` methods*
10. [`copy_cohort`](https://github.com/NGEET/fates/blob/5bb6e3e643b547e3c0dadcbe065e73ffa21b1053/biogeochem/EDCohortDynamicsMod.F90#L1848)
- copies all attributes from one cohort to another empty cohort
- *seems like `cohort` methods*
11. [`count_cohorts`](https://github.com/NGEET/fates/blob/5bb6e3e643b547e3c0dadcbe065e73ffa21b1053/biogeochem/EDCohortDynamicsMod.F90#L2014)
- counts the cohorts within a patch's linked list
- _seems like `patch` methods_
12. [`EvaluateAndCorrectDBH`](https://github.com/NGEET/fates/blob/5bb6e3e643b547e3c0dadcbe065e73ffa21b1053/biogeochem/EDCohortDynamicsMod.F90#L2130)
- corrects DBH of a cohort to be allometrically consistent
- *`cohort` methods?*
13. [`DamageRecovery`](https://github.com/NGEET/fates/blob/5bb6e3e643b547e3c0dadcbe065e73ffa21b1053/biogeochem/EDCohortDynamicsMod.F90#L2221)
- @JessicaNeedham perhaps you can weigh in?

**Comment by JessicaNeedham** [comment]
Thanks @adrifoster this looks great. [`DamageRecovery`](https://github.com/NGEET/fates/blob/5bb6e3e643b547e3c0dadcbe065e73ffa21b1053/biogeochem/EDCohortDynamicsMod.F90#L2221) allows cohorts with crown damage to regrow some of their crowns. It’s mostly a `cohort` method as it assesses the fraction of a cohort that could recover, based on available CNP. But the recovered portion of a cohort is then assigned to a new cohort and has to be inserted into the cohort linked list, so more of a `patch` method? Maybe this routine will need to be split to fit with the new cohort and patch methods?
